pkg/logql/sharding.go

```diff
@@ -55,3 +65,327 @@ func (c ConcatLogSelectorExpr) String() string {
+	return &DownstreamEvaluator{
+		Downstreamer: downstreamer,
+		defaultEvaluator: NewDefaultEvaluator(
+			QuerierFunc(func(_ context.Context, p SelectParams) (iter.EntryIterator, error) {
+				// TODO(owen-d): add metric here, this should never happen.
+				return nil, errors.New("Unimplemented")
+			}),
+			0,
+		),
+	}
+}
+
+// Evaluator returns a StepEvaluator for a given SampleExpr
+func (ev *DownstreamEvaluator) StepEvaluator(
+	ctx context.Context,
+	nextEv Evaluator,
+	expr SampleExpr,
+	params Params,
+) (StepEvaluator, error) {
+	switch e := expr.(type) {
+
+	case DownstreamSampleExpr:
+		// downstream to a querier
+		var shards []astmapper.ShardAnnotation
+		if e.shard != nil {
+			shards = append(shards, *e.shard)
+		}
+		res, err := ev.Exec(ctx, e.SampleExpr, params, shards)
+		if err != nil {
+			return nil, err
+		}
+		return ResultStepEvaluator(res, params)
+
+	case *ConcatSampleExpr:
+		type result struct {
+			stepper StepEvaluator
+			err     error
+		}
+		ctx, cancel := context.WithCancel(ctx)
+		cur := e
+		ch := make(chan result)
+		done := make(chan struct{})
+		count := 0
+
+		for cur != nil {
+			go func(expr SampleExpr) {
+				eval, err := ev.StepEvaluator(ctx, nextEv, expr, params)
+				select {
+				case <-done:
+				case ch <- result{eval, err}:
+				}
+			}(cur.SampleExpr)
+			cur = cur.next
+			count++
+		}
+
+		xs := make([]StepEvaluator, 0, count)
+		cleanup := func() {
+			cancel()           // cancel ctx
+			done <- struct{}{} // send done signal to awaiting goroutines
+			// Close previously opened StepEvaluators
+			for _, x := range xs {
+				x.Close() // close unused StepEvaluators
+			}
+		}
+
+		for i := 0; i < count; i++ {
+			select {
+			case <-ctx.Done():
+				defer cleanup()
+				return nil, ctx.Err()
+			case res := <-ch:
+				if res.err != nil {
+					defer cleanup()
+					return nil, res.err
+				}
+				xs = append(xs, res.stepper)
+			}
+		}
+
+		return ConcatEvaluator(xs)
+
+	default:
+		return ev.defaultEvaluator.StepEvaluator(ctx, nextEv, e, params)
+	}
+}
+
+// Iterator returns the iter.EntryIterator for a given LogSelectorExpr
+func (ev *DownstreamEvaluator) Iterator(
+	ctx context.Context,
+	expr LogSelectorExpr,
+	params Params,
+) (iter.EntryIterator, error) {
+	switch e := expr.(type) {
+	case DownstreamLogSelectorExpr:
+		// downstream to a querier
+		var shards Shards
+		if e.shard != nil {
+			shards = append(shards, *e.shard)
+		}
+		res, err := ev.Exec(ctx, e.LogSelectorExpr, params, shards)
+		if err != nil {
+			return nil, err
+		}
+
+		return ResultIterator(res, params)
+
+	case *ConcatLogSelectorExpr:
+		type result struct {
+			iterator iter.EntryIterator
+			err      error
+		}
+		ctx, cancel := context.WithCancel(ctx)
+		cur := e
+		ch := make(chan result)
+		done := make(chan struct{})
+		count := 0
+
+		for cur != nil {
+			go func(expr LogSelectorExpr) {
+				iterator, err := ev.Iterator(ctx, expr, params)
+				select {
```

Comment:
same here I think it's better to add a warning in case the done swallow the response.